src/ZkFold/Symbolic/Data/Compare.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE DerivingVia          #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module ZkFold.Symbolic.Data.Compare
+  ( Ordering
+  , lt
+  , eq
+  , gt
+  , Ord (..)
+  , (<=)
+  , (<)
+  , (>=)
+  , (>)
+  , min
+  , max
+  ) where
+
+import           Data.Foldable                         (Foldable (..))
+import           Data.Functor.Rep                      (Representable)
+import           GHC.Generics
+import qualified Prelude                               as Haskell
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.VectorSpace
+import           ZkFold.Symbolic.Data.Bool
+import           ZkFold.Symbolic.Data.Container
+
+newtype Ordering a = Ordering a
+  deriving stock
+    ( Haskell.Functor
+    , Haskell.Foldable
+    , Haskell.Traversable
+    )
+deriving via Par1 instance VectorSpace a Ordering
+instance DiscreteField a => Eq a Ordering
+instance (Ring a, Haskell.Eq a) => Haskell.Show (Ordering a) where
+    show (Ordering a) =
+        if a Haskell.== negate one then "<"
+        else if a Haskell.== zero then "="
+        else ">"
+
+lt, eq, gt :: Ring a => Ordering a
+lt = Ordering (negate one)
+eq = Ordering zero
+gt = Ordering one
+
+instance Ring a => Haskell.Semigroup (Ordering a) where
+  Ordering x <> Ordering y = Ordering (x * x * (x - y) + y)
+instance Ring a => Haskell.Monoid (Ordering a) where
+  mempty = eq
+
+class Eq a u => Ord a u where
```

Comment:
For context, see my comment about ByteString's proposed representation.
Do you see how can it fit in here? We'd like to see
```
compare :: ByteString n ArithmeticCircuit a -> ByteString n ArithmeticCircuit a -> Ordering (ArithmeticCircuit a)
```
Do you think making something like
```
newtype Ordering b a = Ordering (b a)
```
and thus separating field elements and computational backend will work? This is how I did it in my branch

Reply:
Instead of using backends, if we have structured multi-outputs, then you can just directly use `ArithmeticCircuit (ByteString n) a` or `ByteString n (ArithmeticCircuit U1 a)`. I don't think backend is necessary.

Reply:
Or perhaps you can instead define a type family like so:
```Haskell
type family Dim (a :: Type) (u :: Type -> Type) :: Natural
type instance Dim a (UInt n) = RegisterCount a n
```
Then you can use either `ArithmeticCircuit (Dim a u) a` or `u (ArithmeticCircuit 1 a)` as you need.

Reply:
If you find a need to add methods to convert back and forth, then you can use a typeclass:

```Haskell
class HasDim a u where
  type family Dim a u :: Natural
  solder :: u (ArithmeticCircuit 1 a) -> ArithmeticCircuit (Dim a u) a
  desolder :: ArithmeticCircuit (Dim a u) a -> u (ArithmeticCircuit 1 a)
```

With structured outputs, `solder` and `desolder` can be implemented once without intermediate list (un)marshalling.

Reply:
Sorry, I still don't see how it all fits together. Can you elaborate further, please? Oh, and keep in mind that my PR focuses ONLY on performance. I totally agree that your approach is beautiful, I just want Symbolic to stop doing redundant computations. 

If we have `ByteString n (ArithmeticCircuit Par1 a)`, then what is the definition of ByteString itself? Does it hold a list (or any other container) of ArithmeticCircuits? If so, then evaluating ByteStrings will involve doing the same computations over and over again because all arithmetic circuits will share the same variables and constraints. 
If we have `ArithmeticCircuit (ByteString n) a`, how do we implement ByteStrings on Zp? Will it be a completely unrelated type?
The implementation of ByteStrings and UInts proposed by me solves both problems. ByteString n ArithmeticCircuit a is just a newtype over ArithmeticCircuit n a (a single circuit with n outputs, which somewhat resembles your second option). ByteString on Zp's is just ByteString n Vector (Zp p) which is just a newtype over Vector n (Zp p).
So we have the same newtype being able to hold both ArithmeticCircuits without any redundand computations and Zp's.

Reply:
> If we have ByteString n (ArithmeticCircuit Par1 a), then what is the definition of ByteString itself?

The definition of `ByteString n` is the same as it is now on the main branch:

```Haskell
newtype ByteString (n :: Natural) a = ByteString [a]
```

> Does it hold a list (or any other container) of ArithmeticCircuits?

`ByteString n` is a list newtype so it can hold a list of circuits or of anything.
`ByteString n (ArithmeticCircuit Par1 a)` holds a list of single-output circuits.

> If so, then evaluating ByteStrings will involve doing the same computations over and over again because all arithmetic circuits will share the same variables and constraints.

That's true, which is why you want to compute things on a multi-output, single-circuit `ArithmeticCircuit (ByteString n) a` instead.

> how do we implement ByteStrings on Zp?

You can directly use `ByteString n (Zp p)` as in the main branch.

> Will it be a completely unrelated type?

It is the exact same symbolic datatype `ByteString n :: Type -> Type`.

> ByteString n ArithmeticCircuit a is just a newtype over ArithmeticCircuit n a (a single circuit with n outputs, which somewhat resembles your second option).

It is equivalent to my second option, which is a single circuit with `n` outputs, `ArithmeticCircuit (ByteString n) a`.

With `UInt`, the second option is a single circuit with `numberOfRegisters a n` outputs, `ArithmeticCircuit (UInt n) a`.

Instead of using a heterogeneous vector type `Vector dim` to contain the outputs of the circuit, which forces you to calculate `dim` at the type level from `n` and `a`, you use homogeneous symbolic datatypes (vector spaces) `ByteString n` or `UInt n` to achieve the same effect.

your proposal is equivalent to my proposal as:
* your `u Vector a` is equivalent to my (and main branch's) `u a`
* your `u ArithmeticCircuit a` is equivalent to my `ArithmeticCircuit u a`
* your multi-output single-circuit `KnownNat n => ArithmeticCircuit n a` is equivalent to my `(VectorSpace a o, Traversable o) => ArithmeticCircuit o a`